core/server/master/src/main/java/alluxio/master/meta/checkconf/ServerConfigurationChecker.java

```diff
@@ -0,0 +1,251 @@
+     * whether errors/warnings exist in masters/workers.
+     */
+    private final Map<Scope, Status> mConfigStatus;
+    /**
+     *  Record the overall status of config check report.
+     *  If any errors (Status = FAILED) exist in Configuration status, the Status here
+     *  will be FAILED, else if any warnings exist, the Status will be WARN.
+     *  If no errors or warnings exist, the Status will be PASSED.
+     */
+    private final Status mReportStatus;
+
+    /**
+     * Creates a new instance of {@link ConfigCheckReport}.
+     */
+    private ConfigCheckReport() {
+      mConfigErrors = new HashMap<>();
+      mConfigWarns = new HashMap<>();
+      mConfigStatus = new HashMap<>();
+      mConfigStatus.put(Scope.MASTER, Status.NOT_STARTED);
+      mConfigStatus.put(Scope.WORKER, Status.NOT_STARTED);
+      mReportStatus = Status.NOT_STARTED;
+    }
+
+    /**
+     * Creates a new instance of {@link ConfigCheckReport}.
+     *
+     * @param configErrors the configuration errors
+     * @param configWarns the configuration warnings
+     * @param configStatus the configuration status
+     * @param reportStatus the overall report status
+     */
+    private ConfigCheckReport(Map<Scope, List<WrongProperty>> configErrors,
+        Map<Scope, List<WrongProperty>> configWarns, Map<Scope, Status> configStatus,
+        Status reportStatus) {
+      mConfigErrors = configErrors;
+      mConfigWarns = configWarns;
+      mConfigStatus = configStatus;
+      mReportStatus = reportStatus;
+    }
+
+    /**
+     * @return a map of configuration errors
+     */
+    public Map<Scope, List<WrongProperty>> getConfigErrors() {
+      return mConfigErrors;
+    }
+
+    /**
+     * @return a map of configuration warnings
+     */
+    public Map<Scope, List<WrongProperty>> getConfigWarns() {
+      return mConfigWarns;
+    }
+
+    /**
+     * @return the status of the configuration checker results
+     */
+    public Map<Scope, Status> getConfigStatus() {
+      return mConfigStatus;
+    }
+
+    /**
+     * @return the overall report status
+     */
+    public Status getReportStatus() {
+      return mReportStatus;
+    }
+  }
+
+  /**
+   * Constructs a new {@link ServerConfigurationChecker}.
+   *
+   * @param masterRecord master configuration record
+   * @param workerRecord worker configuration record
+   */
+  public ServerConfigurationChecker(ServerConfigurationRecord masterRecord,
+      ServerConfigurationRecord workerRecord) {
+    mMasterRecord = masterRecord;
+    mWorkerRecord = workerRecord;
+    mConfigCheckReport = new ConfigCheckReport();
+    mMasterRecord.registerRegenerateReportListener(this::regenerateReport);
+  }
+
+  /**
+   * Checks the server-side configurations and records the check results.
+   */
+  public synchronized void regenerateReport() {
+    // Generate the configuration map from master and worker configuration records
+    Map<PropertyKey, Map<String, List<String>>> confMap = generateConfMap();
```

Comment:
isn't what you need is [`Multimap`](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/Multimap.html)?

Reply:
If I use Multimap, I do make the codes easier here, but the codes will be more difficult in the front end.
The InconsistentProperty will include MultiMap<String, String> value to hostnames field, and we cannot use easy string.join at that time.

When using original Map<String, List<String>>, I can do the following:
```
<% String value = String.format("%s (%s)", entry.getKey(), String.join(", ", entry.getValue()));%>
```

Thanks for providing me a good Multimap option. I think it will be really useful if I do not need to join the values.